tmt/utils/filesystem.py

```
@@ -0,0 +1,332 @@
+        logger.warning(f"Basic copy failed during processing: {e}")
+        return False
+
+    logger.debug(f"Directory copied using basic copy: {copied_count} items processed.")
+    return True
+
+
+def copy_tree(
+    src: Path,
+    dst: Path,
+    logger: 'tmt.log.Logger',
+    workdir_root: Path,
+) -> None:
+    """
+    Copy directory efficiently, trying different strategies.
+
+    Attempts strategies in order:
+    1. Reflinks (copy-on-write)
+    2. Hardlinks with persistent caching for unchanged files
+    3. Basic recursive copy
+
+    Symlinks are always preserved.
+
+    :param src: Source directory path
+    :param dst: Destination directory path
+    :param logger: Logger to use for debug messages
+    :param workdir_root: The root directory for tmt's working files (e.g., /var/tmp/tmt),
+                         used for the hardlink cache.
+    """
+    logger.debug(f"Copying directory tree from '{src}' to '{dst}'")
+
+    # 1. Try reflink copy
+    if _copy_tree_reflinks(src, dst, logger, workdir_root):
+        logger.debug("Copy finished using reflink strategy.")
+        return
+
+    # 2. Try hardlink with cache copy
+    # TODO Implement this!
+    # if _copy_tree_hardlinks(src, dst, logger, workdir_root):
+    #    logger.debug("Copy finished using hardlink/cache strategy.")
+    #    return
+
+    # 3. Fallback to basic copy
+    # This should ideally only be reached if hardlinks failed catastrophically
+    # (e.g., cache dir issue) or if we explicitly add conditions to skip hardlinks.
+    logger.debug("Falling back to basic copy strategy.")
+    if _copy_tree_basic(src, dst, logger, workdir_root):
+        logger.debug("Copy finished using basic copy strategy.")
+        return
+
+    # Should not be reached if _copy_tree_basic always returns True on completion,
+    # unless it encounters an error and returns False.
+    logger.warning(f"All copy strategies failed for '{src}' to '{dst}'.")
+    # Raise exception here?
```

I think we should raise an exception, we were tasked with copying a tree and we failed.

I'm not sure about individual methods and them catching exceptions. The reflink does have a genuine interest in catching the "reflink not supported" issue, and turning it into `False`, but the rest of the errors should probably not result in a fallback. Like running out of disk space. For example:

```python
    except Exception as e:
        logger.warning(f"Basic copy failed during processing: {e}")
        return False
```

I'd argue this should be an exception, e.g. `GeneralError(...) from e`. And I would even say that these methods should not bother with exceptions unless they can do something better than logging and reraising them, I'd think that `copy_tree()` could wrap calls with `try` and `except` and wrap whatever is raised in individual methods with `GeneralError("Failed to copy tree") from exc` or something like that.